### DISCUSSION TITLE
Right way to us FileChannel?

### DISCUSSION THREAD

**Discussion by Anivie** [netty/netty#12726] [Q&A]
I am tring to build a netty file server,it is great for me to use TCP for file transfer, but I can't get the file transfer progress when I use FileChannel, what is the correct way to use it?

This is My Server code:
```kotlin
RandomAccessFile("C:\\Users\\Anivie\\Desktop\\demo.pdf","r").run {
   out += ctx.alloc().buffer().apply {
      writeInt(length().toInt())
      writeBytes(channel,0,length().toInt())
   }
}
```

Client code:
```kotlin
RandomAccessFile("C:\\Users\\Anivie\\Desktop\\demo-1.pdf","rw").run {
   msg.readBytes(channel,0, msg.readInt())
 }
```

I can't find the correct way to transfer files on it, there are only examples of using http in the netty example, is there any documentation or example for using TCP?

**Comment by chrisvest** [comment]
You should do some kind of chunking, reading the file in smaller pieces, then track the progress on how many pieces you've sent vs how many are left to send.

This is also a requirement to support large files, since buffers cannot be bigger than 2 GiB, and you cannot move more data than that in a single IO.

**Reply by Anivie** [reply]
Yes,this is how i am working on it,but if the file we are trying to transfer is very smaller,it will not working great.Is any better way to process it?
